### ISSUE TITLE
unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.

### ISSUE THREAD

**Issue by klokare** [htm-community/htm.core#878]
running the `htm/advanced/examples/thing_classification` experiment, `l2l4l6_experiment.py`, fails with the following message:

> RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.

Full stack trace below.

Installed on MacOS 10.15.6 (Catalina) using conda with Python 3.8.5. Tested installation with `htm/examples/sp/hello_sp.py` and `htm/examples/mnist.py` and both work fine. 

I have been poking around and traced the error back to RegionImplFactory.cpp:214. From that file, it looks like I need to add some call to `addRegionType` but I do not see an example of how to do that from the Python side given that the `RawSensor` and `RawValue` types are defined only in Python.

```
(htmcore) klokare@sterling thing_classification % python l2l4l6_experiment.py 
Accuracy/cells_per_axis10.0 : 0
Accuracy/cells_per_axis20.0 : 0
Accuracy/cells_per_axis30.0 : 0
Traceback (most recent call last):
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
Traceback (most recent call last):
Accuracy/cells_per_axis40.0 : 0
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
Accuracy/cells_per_axis50.0 : 0
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
Traceback (most recent call last):
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
Accuracy/cells_per_axis60.0 : 0
Traceback (most recent call last):
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
Traceback (most recent call last):
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(n  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
Traceback (most recent call last):
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
Accuracy/cells_per_axis70.0 : 0
Accuracy/cells_per_axis80.0 : 0
Accuracy/cells_per_axis90.0 : 0
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in re  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
set
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
etwork=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
Accuracy/cells_per_axis100.0 : 0
Traceback (most recent call last):
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/klokare/miniconda3/envs/htmcore/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/klokare/miniconda3/envs/htmcore/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 30, in mp_runrep
    return PyExperimentSuite.run_rep(*args)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 638, in run_rep
    self.reset(params, rep)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/examples/thing_classification/l2l4l6_experiment.py", line 146, in reset
    self.network = L246aNetwork(numColumns=1,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/logging_decorator.py", line 84, in _fn
    return fn(instance, *args, **kwargs)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 438, in __init__
    self.network = createMultipleL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 359, in createMultipleL246aLocationColumn
    network = createL246aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 245, in createL246aLocationColumn
    network = createL4L6aLocationColumn(network=network,
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/frameworks/location/location_network_creation.py", line 126, in createL4L6aLocationColumn
    network.addRegion(sensorInputName, "py.RawSensor", json.dumps({"outputWidth": columnCount}))
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "l2l4l6_experiment.py", line 247, in <module>
    suite.start()
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 576, in start
    self.do_experiment(paramlist)
  File "/Users/klokare/Code/reference/htm.core/build/Release/distr/src/htm/advanced/support/expsuite.py", line 609, in do_experiment
    pool.map(mp_runrep, explist)
  File "/Users/klokare/miniconda3/envs/htmcore/lib/python3.8/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/Users/klokare/miniconda3/envs/htmcore/lib/python3.8/multiprocessing/pool.py", line 771, in get
    raise self._value
RuntimeError: Exception: RegionImplFactory.cpp(213) message: getSpec() -- unknown node type: 'py.RawSensor'.  Custom node types must be registed before they can be used.
```

**Comment by dkeeney** [comment]
In C++ the RawInput region type is an internal region created automatically if you declare a Link with the source "INPUT".  This is so that you can use the function net.setInputData(sourceName, data) from an app to provide data to the input.

However, this appears to be something different.  All python region types will be prefixed with py. to avoid name collisions.  So what you need to do is call net.registerPyRegion(module, classname) to register a region written in python.  This will take the classname, prepend py. to it, and that is the name to use when calling addRegion().

```
 net = Network()
 net.registerPyRegion("htm/advanced/examples/thing_classification/RawSensor", "RawSensor")
 net.addRegion("sensor", "py.RawSensor")
```
I probably have the module path wrong but I think you can see the pattern.

I know... we need better documentation :)

**Comment by klokare** [comment]
Thanks, @dkeeney. No worries on the documentation. I can follow (most of) the code and this is how I learn best. I see that the registering for the experiment happens in the `htm/advanced/support/register_regions.py` methods. If I step through the code in a debugger I see the RawSensor module being registered. The call is

```python
    if moduleName is None:
        # the region is located in the regions/ directory
        moduleName = "htm.advanced.regions." + regionTypeName
    # Add new region class to the network.
    module = __import__(moduleName, {},{}, [regionTypeName], 0)
    Network.registerPyRegion(module.__name__, regionTypeName)
```

which equates to the last line being 

```python
Network.registerPyRegion('htm.advanced.regions.RawSensor', 'RawSensor')
```

Ismpecting the `module` variable looks OK. That is, the `__import__` seems to have worked. Is the issue just needing to change the `registerPyRegion` call to use slashes instead of dots?

**Comment by dkeeney** [comment]
> needing to change the registerPyRegion call to use slashes instead of dots?

Oh, my bad.  Those should be dots.  (You can tell I am a C++ programmer and not a Python programmer).

**Comment by klokare** [comment]
The problem seems to be in using Python 3.8 instead of 3.7. I tested a clean install and build from scratch with both versions on Mac and Linux and the Python 3.7 works without issue and Python 3.8 has problems, though different ones depending on the platform. As my issue is solved with Python 3.7 I am marking this issue closed. Thanks, @dkeeney for the help. It may not have solved the Python 3.8 issue but I learned a good bit more about the code.

NOTE: I also am changing the title of this issue. The problem is correct but the error shows RawInput instead of RawSensor. I mistakenly copied the output of one of my failed attempts at troubleshooting. The same error occurs with RawSensor, though.
